cluster-autoscaler/core/utils_test.go

```diff
@@ -379,70 +379,190 @@ func TestFilterSchedulablePodsForNode(t *testing.T) {
+	SetNodeReadyState(ready1, true, time.Now())
+	ready2 := BuildTestNode("n2", 2000, 2000)
+	SetNodeReadyState(ready2, true, time.Now())
+	unready3 := BuildTestNode("n3", 3000, 3000)
+	SetNodeReadyState(unready3, false, time.Now())
+	unready4 := BuildTestNode("n4", 4000, 4000)
+	SetNodeReadyState(unready4, false, time.Now())
+	ready5 := BuildTestNode("n5", 5000, 5000)
+	SetNodeReadyState(ready5, true, time.Now())
+	ready6 := BuildTestNode("n6", 6000, 6000)
+	SetNodeReadyState(ready6, true, time.Now())
+
+	tn := BuildTestNode("tn", 10000, 10000)
+	tni := schedulercache.NewNodeInfo()
+	tni.SetNode(tn)
+
+	lastDeletedGroup := ""
+	onDeleteGroup := func(id string) error {
+		lastDeletedGroup = id
+		return nil
+	}
+
+	// Cloud provider with TemplateNodeInfo implemented.
+	provider1 := testprovider.NewTestAutoprovisioningCloudProvider(
+		nil, nil, nil, onDeleteGroup, nil,
+		map[string]*schedulercache.NodeInfo{"ng3": tni, "ng4": tni})
+	provider1.AddNodeGroup("ng1", 1, 10, 1) // Nodegroup with ready node.
+	provider1.AddNode("ng1", ready1)
+	provider1.AddNodeGroup("ng2", 1, 10, 1) // Nodegroup with ready and unready node.
+	provider1.AddNode("ng2", ready2)
+	provider1.AddNode("ng2", unready3)
+	provider1.AddNodeGroup("ng3", 1, 10, 1) // Nodegroup with unready node (and 1 previously ready node).
+	provider1.AddNode("ng3", unready4)
+	provider1.AddNode("ng3", ready5)
+	provider1.AddNodeGroup("ng4", 0, 1000, 0) // Nodegroup without nodes (and 1 previously ready node).
+	provider1.AddNode("ng4", ready6)
+
+	podLister := kube_util.NewTestPodLister([]*apiv1.Pod{})
+	registry := kube_util.NewListerRegistry(nil, nil, podLister, nil, nil, nil, nil, nil, nil, nil)
+
+	predicateChecker := simulator.NewTestPredicateChecker()
+
+	nodeInfoCache := make(map[string]*schedulercache.NodeInfo)
+
+	// Fill cache
+	res, err := getNodeInfosForGroups([]*apiv1.Node{unready4, unready3, ready2, ready1}, nodeInfoCache,
+		provider1, registry, []*appsv1.DaemonSet{}, predicateChecker)
+	assert.NoError(t, err)
+	// Check results
+	assert.Equal(t, 4, len(res))
+	info, found := res["ng1"]
+	assert.True(t, found)
+	assertEqualNodeCapacities(t, ready1, info.Node())
+	info, found = res["ng2"]
+	assert.True(t, found)
+	assertEqualNodeCapacities(t, ready2, info.Node())
+	info, found = res["ng3"]
+	assert.True(t, found)
+	assertEqualNodeCapacities(t, tn, info.Node())
+	info, found = res["ng4"]
+	assert.True(t, found)
+	assertEqualNodeCapacities(t, tn, info.Node())
+	// Check cache
+	cachedInfo, found := nodeInfoCache["ng1"]
+	assert.True(t, found)
+	assertEqualNodeCapacities(t, ready1, cachedInfo.Node())
+	cachedInfo, found = nodeInfoCache["ng2"]
+	assert.True(t, found)
+	assertEqualNodeCapacities(t, ready2, cachedInfo.Node())
+	cachedInfo, found = nodeInfoCache["ng3"]
+	assert.False(t, found)
+	cachedInfo, found = nodeInfoCache["ng4"]
+	assert.False(t, found)
+
+	// Invalidate part of cache in two different ways
+	provider1.DeleteNodeGroup("ng1")
```

Comment:
Doesn't the below call (nodeGroup.Delete()) also call this internally?

Reply:
It does, with small differences. Those two different way do delete a group test more those mocks than anything else. I think that it is still beneficial to leave it.

Reply:
Not sure I'm following - Delete() doesn't call DeleteNodeGroup() if and only if callback failed; we aren't checking any error in the below call, do we expect it to fail?